### PR TITLE
schedule: refactor region scatter (#3451)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #3451 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In pd repo:
git pr https://github.com/tikv/pd/pull/3555
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/pd.git pr/3555:release-4.0-7ef4f4273cdd
```

---

### What problem does this PR solve?
ref https://github.com/tikv/pd/issues/3450
Also the code in region scatter is too complex to maintain.

### What is changed and how it works?
refactor the region scatter to make it simple and fix the reference bug.

After this pr, the result was like following: 

```sh
mysql> create table t(id int);
Query OK, 0 rows affected (0.09 sec)

mysql> SPLIT TABLE t BETWEEN (0) AND (1000000000) REGIONS 200;                                +--------------------+----------------------+
| TOTAL_SPLIT_REGION | SCATTER_FINISH_RATIO |
+--------------------+----------------------+
|                199 |                    1 |
+--------------------+----------------------+
1 row in set (11.89 sec)

mysql> select count(s.region_id) cnt, s.index_name, p.store_id from INFORMATION_SCHEMA.TIKV_REGION_STATUS s join INFORMATION_SCHEMA.tikv_region_peers p on s.region_id = p.region_id where s.table_name = 't' and p.is_leader = 0 group by index_name, p.store_id order by index_name,cnt desc;
+-----+------------+----------+
| cnt | index_name | store_id |
+-----+------------+----------+
|  70 | NULL       |        1 |
|  67 | NULL       |        4 |
|  67 | NULL       |        6 |
|  66 | NULL       |        7 |
|  65 | NULL       |        5 |
|  65 | NULL       |        8 |
+-----+------------+----------+
6 rows in set (0.02 sec)
```

For the detail test result, please see:
https://docs.google.com/document/d/1cGZU2DCm0UuIueqAQP8TGqTivyavCvB_QiwIzqvP_rA/edit?usp=sharing

### Check List

Tests

- Unit test

### Release note

* No release note

### CODE REVIEW

File: server/schedule/region_scatterer.go [hunk 1]
```diff
@@ -207,90 +276,87 @@ func (r *RegionScatterer) scatterRegion(region *core.RegionInfo, group string) *
...
+		regionStores := make([]*core.StoreInfo, 0)
+		for _, peer := range region.GetPeers() {
+			store := r.cluster.GetStore(peer.GetStoreId())
+			regionStores = append(regionStores, store)
+		}
 		scoreGuard = filter.NewDistinctScoreFilter(r.name, r.cluster.GetLocationLabels(), regionStores, sourceStore)
 	}
-
-	candidates := make([]*core.StoreInfo, 0, len(stores))
+	//scoreGuard := filter.NewRuleFitFilter(r.name, r.cluster, region, sourceStore.GetID())
```
Comment (nolouch) [inline server/schedule/region_scatterer.go:321]: remove this line?

File: server/schedule/region_scatterer.go [hunk 2]
```diff
@@ -136,67 +120,152 @@ type engineContext struct {
...
 	}
+	return ops, nil
 }
 
-// Scatter relocates the region. If the group is defined, the regions's leader with the same group would be scattered
+// Scatter relocates the region. If the group is defined, the regions' leader with the same group would be scattered
 // in a group level instead of cluster level.
 func (r *RegionScatterer) Scatter(region *core.RegionInfo, group string) (*operator.Operator, error) {
 	if !opt.IsRegionReplicated(r.cluster, region) {
+		//r.cluster.AddSuspectRegions(region.GetID())
```
Comment (nolouch) [inline server/schedule/region_scatterer.go:228]: remove this line?

**Review by JmPotato — APPROVED** [review]

File: pkg/typeutil/comparison.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2016 TiKV Project Authors.
```
Comment (rleungx) [inline pkg/typeutil/comparison.go:1]:
```suggestion
// Copyright 2021 TiKV Project Authors.
```

File: server/schedule/region_scatterer_test.go [hunk 4]
```diff
@@ -105,15 +114,16 @@ func (s *testScatterRegionSuite) scatter(c *C, numStores, numRegions uint64, use
...
 			countPeers[peer.GetStoreId()]++
-			if peer.GetId() == region.GetLeader().GetId() {
+			if peer.GetStoreId() == leaderStoreID {
 				countLeader[peer.GetStoreId()]++
 			}
 		}
 	}
 
-	// Each store should have the same number of peers.
+	//Each store should have the same number of peers.
```
Comment (rleungx) [inline server/schedule/region_scatterer_test.go:126]:
```suggestion
	// Each store should have the same number of peers.
```

**Pull request merged**